test/jdk/javax/swing/SwingWorker/TestDoneBeforeDoInBackground.java

```diff
@@ -0,0 +1,90 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+/*
+ * @test
+ * @bug 8081474
+ * @summary  Verifies if SwingWorker calls 'done'
+ *           before the 'doInBackground' is finished
+ * @run main TestDoneBeforeDoInBackground
+ */
+import javax.swing.SwingWorker;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+public class TestDoneBeforeDoInBackground {
+
+    static boolean doInBackground = false;
+    static final int WAIT_TIME = 200;
+    private static final long CLEANUP_TIME = 3000;
+
+    public static void main(String[] args) throws InterruptedException {
+        SwingWorker<String, String> worker =
+            new SwingWorker<String, String>() {
+            @Override
+            protected String doInBackground() throws Exception {
+                try {
+                    while (!Thread.currentThread().isInterrupted()) {
+                        System.out.println("Working...");
+                        Thread.sleep(WAIT_TIME);
+                    }
+                } catch (InterruptedException ex) {
+                    System.out.println("Got interrupted!");
+                }
+                try {
+                    System.out.println("Cleaning up");
+                    Thread.sleep(CLEANUP_TIME);
+                    System.out.println("Done cleaning");
+                    doInBackground = true;
+                } catch (InterruptedException ex) {
+                    System.out.println("Got interrupted second time!");
+                }
+                return null;
+            }
+
+            @Override
+            protected void done() {
+                if (!doInBackground) {
+                    throw new RuntimeException("done called before doInBackground");
+                }
+                System.out.println("Done");
+            }
+        };
+
+        worker.execute();
+        Thread.sleep(WAIT_TIME * 3);
+
+        final long start = System.currentTimeMillis();
+        worker.cancel(true);
+        final long end = System.currentTimeMillis();
+
+        if ((end - start) > 100) {
+            throw new RuntimeException("Cancel took too long: "
+                                       + ((end - start) / 1000.0d) + " s");
+        }
+        if (doInBackground && worker.getState() != SwingWorker.StateValue.DONE) {
+            throw new RuntimeException("doInBackground is finished " +
+                                       " but State is not DONE");
+        }
```

Comment:
I think you should wait for cleanup to finish and ensure that `doInBackground` is set to `true` and the state is `DONE`.

If you don't wait until cleanup wait completes, the test does not test anything because jtreg will shutdown the JVM as soon main thread exists, which will happen before `done` gets called.

Reply:
Dont think so...If `doInBackground `is true it should mean cleanup is done as `doInBackground ` is false by default

Reply:
That's the thing. If `cancel` doesn't wait until `doInBackground` completes, which *it shouldn't do*, then `cancel` returns quickly. Therefore the main thread reaches this condition before the “cleanup” in `doInBackground` finishes, as such the test should always fail.

Do I miss anything?

Reply:
ok..understood...modified...